### ISSUE TITLE
make_fcontext segfault

### ISSUE THREAD

**Issue by adammoody** [LLNL/UnifyFS#623]
Running a particular HDF test case with two client procs on two nodes causes the server to hit a segfault.  In this case, the second process is executing lots of reads, which are ultimately handled by the server on the first node.  The chunk_read_request_rpc handler eventually throws a segfault.  This failure seems to be racy, since it's not the same iteration of the rpc call each time.

I can see that it happens somewhere on the 220-320th read call.  Not sure how many underlying rpcs that ends up launching though.

It appears to come from this source line:
https://github.com/pmodels/argobots/blob/9f08f49cbe21d7526438dbc6711ec10d9bd0717b/src/arch/fcontext/make_ppc64_sysv_elf_gas.S#L179

The call stack at the point of the failure looks like:
```
Stack Trace
     make_fcontext,       FP=2000023e7980
     ABT_thread_create,   FP=2000023e7980
C    _handler_for_chunk_read_request_rpc, FP=2000023e7ab0
C    hg_core_rpc_cb,      FP=2000023e7b20
C    hg_core_process,     FP=2000023e7c70
C    hg_core_trigger_entry, FP=2000023e7c70
C    hg_core_trigger,     FP=2000023e7c70
C    HG_Core_trigger,     FP=2000023e7ce0
C    HG_Trigger,          FP=2000023e7d50
     __margo_hg_progress_fn, FP=2000023e7d50
     ABTD_thread_func_wrapper_thread, FP=2000023e7e10
     make_fcontext,       FP=2000023e7e40
```
The segfault itself happens on the ``std r5,520(r3)`` instruction below:
```
  ::  make_fcontext:          0x3c4c0003  addis     rtoc,r12,3
  ::      0x20000023b33c:     0x3842cbc8  subi      rtoc,rtoc,13368
  ::      0x20000023b340:     0x7cc802a6  mflr      r6
  ::      0x20000023b344:     0x786306e4  rldicr    r3,r3,0,59
  ::      0x20000023b348:     0x3863fdb0  subi      r3,r3,592
  =>      0x20000023b34c:     0xf8a30208  std       r5,520(r3)
  ::      0x20000023b350:     0x7c0802a6  mflr      r0
  ::      0x20000023b354:     0x48000005  bl        0x20000023b358
  ::      0x20000023b358:     0x7c8802a6  mflr      r4
  ::      0x20000023b35c:     0x38840018  addi      r4,r4,24
  ::      0x20000023b360:     0x7c0803a6  mtlr      r0
  ::      0x20000023b364:     0xf8830200  std       r4,512(r3)
  ::      0x20000023b368:     0x7cc803a6  mtlr      r6
  ::      0x20000023b36c:     0x4e800020  blr
  ::  finish:                 0x7c0802a6  mflr      r0
```
The values in the stack frame (for R5 and R3):
```
Stack Frame
- - - - - - - - - - - - - - - - - - - - - - - - - - - -
Registers for the frame:

    R0: 0x000000001004e220 (268755488)
    SP: 0x00002000023e7980 (35184409737600)
  RTOC: 0x0000200000267f00 (35184374611712)
    R3: 0x0000000000003cb0 (15536)
    R4: 0x0000000000003f00 (16128)
    R5: 0x000020000023a6d0 (35184374425296)
    R6: 0x0000200000233ac8 (35184374397640)
    R7: 0x0000000000000000 (0)
    R8: 0x0000000000000000 (0)
    R9: 0x0000000000003f00 (16128)
   R10: 0x0000200040027738 (35185445992248)
   R11: 0x0000000010c29e04 (281189892)
   R12: 0x000020000001bb50 (35184372202320)
   R13: 0x00002000184168b0 (35184779028656)
   R14: 0x0000000000000000 (0)
   R15: 0x0000000010c22840 (281159744)
   R16: 0x0000000010e071c0 (283144640)
   R17: 0x0000000010e07338 (283145016)
   R18: 0x0000000010e0d590 (283170192)
   R19: 0x0000000000000001 (1)
   R20: 0x0000000010c228f0 (281159920)
   R21: 0x0000000000000002 (2)
   R22: 0x0000000010e07a20 (283146784)
   R23: 0x000000001004e088 (268755080)
   R24: 0x0000000000000000 (0)
   R25: 0x0000000000000000 (0)
   R26: 0x0000000000003f00 (16128)
   R27: 0x000020001c000900 (35184841853184)
   R28: 0x0000000010c228a0 (281159840)
   R29: 0x0000000010e0d880 (283170944)
   R30: 0x0000000000000000 (0)
   R31: 0x0000200040027680 (35185445992064)
  INUM: 0x0000000000000000 (0)
    PC: 0x20000023b34c (make_fcontext+0x14)
```
My PowerPC assembly is rusty, but I think that's trying to store the value of R5 at memory address (R3)+520 = 15536 + 520.  Looking at the code, I think R3 should be pointing to some location in the thread stack.

**Comment by adammoody** [comment]
Trying to track that back, I can see at least one place where ``make_fcontext`` is called is here:
https://github.com/pmodels/argobots/blob/9b0110b5fcec86bdac0d228bda45a02b8dd90bcc/src/include/abtd_fcontext.h#L30
Which is called from:
https://github.com/pmodels/argobots/blob/9b0110b5fcec86bdac0d228bda45a02b8dd90bcc/src/include/abtd_thread.h#L35
From:
https://github.com/pmodels/argobots/blob/9b0110b5fcec86bdac0d228bda45a02b8dd90bcc/src/include/abtd_thread.h#L47
From:
https://github.com/pmodels/argobots/blob/9b0110b5fcec86bdac0d228bda45a02b8dd90bcc/src/thread.c#L1441
From:
https://github.com/pmodels/argobots/blob/9b0110b5fcec86bdac0d228bda45a02b8dd90bcc/src/thread.c#L65
And then from:
https://xgitlab.cels.anl.gov/sds/margo/-/blob/v0.4.3/include/margo.h#L944

**Comment by adammoody** [comment]
My guess is that the (stack + stack size) computation may have gone bad for some reason, but I don't know why.  Maybe just exhausted the stack?

**Comment by adammoody** [comment]
A ``cat /proc/<pid>/maps`` of the server process had this entry:
```
7ffffffd0000-800000000000 rw-p 00000000 00:00 0                          [stack]
```

**Comment by adammoody** [comment]
@MichaelBrim , did we get to the bottom of the IOR read segfaults?

I'm wondering if @clmendes 's HDF test case might be tripping on it, too.

A second problem here is that our UnifyFS client is returning a valid read with 0 data, instead of an error, when the the server has crashed.

**Comment by MichaelBrim** [comment]
I haven't had a chance to go back and look into the IOR segfaults yet. Could be related, since they were both in a reading phase.

**Comment by adammoody** [comment]
Most of the calls to ``make_fcontext`` that work leading up to the point of the segfault have address values like ``0x0000200000******``.

**Comment by adammoody** [comment]
Oh, argobots is allocating its thread stacks in heap, not the stack, so nevermind about the stack region above.  I'm still trying to figure out why we end up with a bad memory address for the thread stack.

**Comment by adammoody** [comment]
@clmendes , I’m trying to reproduce this segfault with a debug build of argobots so I can figure out what’s gone wrong.  To do that, we need to modify the bootstrap script and then rebuild argobots.

In the bootstrap.sh, replace these two lines in the argobots section:
```
./autogen.sh && CC=gcc ./configure --prefix="$INSTALL_DIR"
make -j $make_nproc && make install
```
with these three lines:
```
./autogen.sh && CC=gcc ./configure CFLAGS=”-g -O0” --prefix="$INSTALL_DIR" --enable-fast=O0
make clean
make V=1 VERBOSE=1 -j $make_nproc && make install
```
There are several changes there.  It may be easiest to cut-and-paste those three lines in.  You could just comment out the existing two lines.

Can you make those changes and then rerun bootstrap?

If we can get the problem to reproduce with that build, we should be able to trace through the argobots code with TotalView.

**Comment by adammoody** [comment]
For future reference, this page has some good tips for debugging and collective performance profiles of margo/argobots: https://github.com/mochi-hpc/mochi-margo/blob/main/doc/debugging.md

**Comment by adammoody** [comment]
To try to simplify, I've written another reproducer where processes interleave writing blocks of 1000 bytes, and each writes about 10,000 such blocks.  Each process then reads back the entire file, also reading in chunks of 1000 bytes.  Allowing just rank 0 to execute the reads, I hit a segfault with the following stack trace:
```
Stack Trace
C    extent_tree_rdlock,  FP=2000680009d0
C    extent_tree_get_chunk_list, FP=200068000a60
C    unifyfs_inode_get_extent_chunks, FP=200068001b30
C    unifyfs_inode_resolve_extent_chunks, FP=200068002c80
C    find_extents_rpc,    FP=200068003e50
C    _wrapper_for_find_extents_rpc, FP=200068003ee0
C    ABTD_thread_func_wrapper_thread, FP=200068003f40
     make_fcontext,       FP=200068003f40
C    extent_tree_get_chunk_list, FP=000090
```
This was interesting, because the server process was blocked on a pthread rdlock in ``extent_tree_get_chunk_list``.  Then it seems a user-level thread started up to service a `` find_extents_rpc`` request, which called the same ``extent_tree_get_chunk_list`` and eventually the same pthread read lock.

It seems calling that lock a second time caused the segfault?

Maybe we need to include the argobots locks?

Still just guessing on this.

**Comment by adammoody** [comment]
Move the above to its own issue in case it may be a separate problem https://github.com/LLNL/UnifyFS/issues/624

**Comment by adammoody** [comment]
@clmendes , since the problem doesn't reproduce with the build using those new flags, let's try to back some of them out to see if we're lucky.  Can you try again with just the following configure line:
```
./autogen.sh && CC=gcc ./configure CFLAGS=”-g -O0” --prefix="$INSTALL_DIR" --enable-fast=O0
```

**Comment by clmendes** [comment]
@adammoody , I rebuilt it with the line you suggested in bootstrap.sh, and it still works! So, at this point, the difference in bootstrap.sh is:

```
[mendes3@lassen708:UnifyFS]$ diff bootstrap.sh.ORIGINAL  bootstrap.sh
174c174,175
< ./autogen.sh && CC=gcc ./configure --prefix="$INSTALL_DIR"
---
> #CLM ./autogen.sh && CC=gcc ./configure --prefix="$INSTALL_DIR"
> ./autogen.sh && CC=gcc ./configure CFLAGS="-g -O0" --prefix="$INSTALL_DIR" --enable-fast=O0
[mendes3@lassen708:UnifyFS]$
```
Thus, differences are only the CFLAGS and the enable-fast. I will retry without the second (will use only CFLAGS) and see what happens.

**Comment by clmendes** [comment]
@adammoody , using only the CFLAGS addition in the argobots build (i.e. removing --enable-fast=O0) the code fails in the Verification phase again!

So, it looks like we need --enable-fast=O0 in the argobots build for the app to work correctly. Does that make any sense?

**Comment by clmendes** [comment]
@adammoody , I just confirmed with a test on Lassen: adding the --enable-fast=O0 to the argobots build line in bootstrap.sh, the calloc test works!

**Comment by clmendes** [comment]
@adammoody , @CamStan : Just to confirm about this, I did a fresh download of Unify on Lassen, and made a single change to bootstrap.sh: added the --enable-fast=O0 to the argobots build line in bootstrap.sh. Then built Unify, with no debug support. Tested it with calloc, and it works fine! The execution took 32 seconds, which is much closer to the non-Unify execution (14 seconds), and the produced HDF5 data files match.

**Comment by adammoody** [comment]
@clmeneds , thanks for narrowing this down.  While we're at it, can you also try ``--enable-fast=01``?

**Comment by adammoody** [comment]
> 
> 
> @adammoody , using only the CFLAGS addition in the argobots build (i.e. removing --enable-fast=O0) the code fails in the Verification phase again!
> 
> So, it looks like we need --enable-fast=O0 in the argobots build for the app to work correctly. Does that make any sense?

There are a class of bugs where the compiler level may expose or hide them, and we're up against one of those here.  It could be something like a race condition (timing), memory corruption (memory layout), or even compiler bugs (far less likely, but I've seen them).  It makes debugging much harder, since we can't see the source and variable values under a debugger, but we'll figure it out.  We might have to start making tweaks to the argobots code for more clues.

**Comment by clmendes** [comment]
@adammoody , it took me a while (too many meetings for a Friday!), but I just tested --enable-fast=O1, and with this it will hang! The job is killed after 20 minutes in the execution.

So, what exactly will --enable-fast do? I tried to google it, but didn't find an explanation yet.

Also, what's the difference between --enable-fast=O0 and CFLAGS=-g -O0   ?

**Comment by adammoody** [comment]
Thanks @clmendes, this option is defined in the argobots configure script:
https://github.com/pmodels/argobots/blob/9f08f49cbe21d7526438dbc6711ec10d9bd0717b/configure.ac#L107-L117
I had to track that down because it's what really sets the ``-O`` flag during the compile step.  Setting ``CFLAGS="-g -O0"`` without setting ``--enable-fast`` gives compile lines like ``-g -O0 -O2``, and then that trailing ``-O2`` overrides the ``-O`` from ``CFLAGS``.

Working with ``-O1`` and a hang might provide some more clues.  I can give that a try next week.  Beyond that, I think we'll next start changing the argobots code and see what we can find.

Can you send the build directory to me in an email?

**Comment by clmendes** [comment]
@adammoody , here's the info on my two versions of the test, one with --enable-fast=O0 (which works) and another with --enable-fast=O1 (which hangs, and the job allocation expires). For both, I'm building Unify and the app with debug support, i.e. -g -O0, everything under LASSEN.

a) Version with --enable-fast=O0:   works OK
Unify Build:  /g/g12/mendes3/LASSEN/UnifyFS-618DD/UnifyFS/
App build:  /usr/workspace/mendes3/LASSEN/CELSO-618DD/
App Execution: /usr/workspace/mendes3/LASSEN/CELSO-618DD/UNIFY/
Execution script:  job.uni  -- Job Output: myoutput.txt  -- Execution stdout: out

b) Version with --enable-fast=O1:  execution hangs
Unify Build:  /g/g12/mendes3/LASSEN/UnifyFS-618DE/UnifyFS/
App build:  /usr/workspace/mendes3/LASSEN/CELSO-618DE/
App Execution: /usr/workspace/mendes3/LASSEN/CELSO-618DE/UNIFY/
Execution script:  job.uni  -- Job Output: myoutput.txt  -- Execution stdout: out

**Comment by adammoody** [comment]
@clmendes , do you still have ``CFLAGS="-g -O0"`` set for the ``--enable-fast=O1`` build?

If not, can you try the ``--enable-fast=O1`` build again with ``CFLAGS="-g -O0"``.  Not having -g might be why I'm still not seeing the source.

**Comment by clmendes** [comment]
@adammoody no, I did not have the CFLAGS in my --enable-fast=-O1 build. I only had the -g for the Unify build itself, not for argobots in bootstratp.sh.

I will do it, after I get rid of a few urgent things here, and will notify you.

**Comment by clmendes** [comment]
@adammoody , I repeated the calloc test as you suggested, and it still hangs. This is the configuration on LASSEN:

* Unify build at  /g/g12/mendes3/LASSEN/UnifyFS-618DF/UnifyFS/
In particular, file bootstrap.sh has for its line 175 the following:
 ./autogen.sh && CC=gcc ./configure CFLAGS="-g -O0" --prefix="$INSTALL_DIR" --enable-fast=O1
and then Unify is built with -g and -O0 too.

* App built at  /usr/workspace/mendes3/LASSEN/CELSO-618DF/
and executed at /usr/workspace/mendes3/LASSEN/CELSO-618DF/UNIFY/
via script in job.uni

**Comment by clmendes** [comment]
And for the records, while I was at it, I also rebuilt everything (Unify and app) using gcc/8.3.1. I did NOT build with debug support, and did not modify the original bootstrap.sh. The result was that the program ran, but the data verification failed again, as we had seen before.

* Unify build:  /g/g12/mendes3/LASSEN/UnifyFS-618G/UnifyFS/

* App build:  /usr/workspace/mendes3/LASSEN/CELSO-618G/
Execution script:  /usr/workspace/mendes3/LASSEN/CELSO-618G/UNIFY/job.uni
Execution output:  /usr/workspace/mendes3/LASSEN/CELSO-618G/UNIFY/out

**Comment by adammoody** [comment]
Thanks @clmendes .  That helped pushing things along.  I can now see the source, and a lot more of the stack.  I can see some of the variable values, but it's hard to trust them with this optimization level.
```
Stack Trace
     make_fcontext,       FP=2000024c7980
C    ABTD_thread_context_make, FP=2000024c7a10
C    ABTDI_thread_context_create, FP=2000024c7a10
C    ABTD_thread_context_create_thread, FP=2000024c7a10
C    ABTI_thread_create_internal, FP=2000024c7a10
C    ABT_thread_create,   FP=2000024c7a10
C    _handler_for_chunk_read_request_rpc, FP=2000024c7ab0
C    hg_core_rpc_cb,      FP=2000024c7b20
C    hg_core_process,     FP=2000024c7c70
C    hg_core_trigger_entry, FP=2000024c7c70
C    hg_core_trigger,     FP=2000024c7c70
C    HG_Core_trigger,     FP=2000024c7ce0
C    HG_Trigger,          FP=2000024c7d50
C    __margo_hg_progress_fn, FP=2000024c7e10
C    ABTD_thread_func_wrapper_thread, FP=2000024c7e40
     make_fcontext,       FP=2000024c7e40
```
That's the path I was guessing it was taking, but we can see that spelled out now.

Backing up a few frames, I see this code where ``p_stacktop`` is computed:
```
  21  static inline int ABTDI_thread_context_create(ABTD_thread_context *p_link,
  22                                                void (*f_wrapper)(void *),
  23                                                void (*f_thread)(void *),
  24                                                void *p_arg, size_t stacksize,
  25                                                void *p_stack,
  26                                                ABTD_thread_context *p_newctx)
  27  {
  28      int abt_errno = ABT_SUCCESS;
  29      void *p_stacktop;
  30
  31      /* ABTD_thread_context_make uses the top address of stack.
  32         Note that the parameter, p_stack, points to the bottom of stack. */
  33      p_stacktop = (void *)(((char *)p_stack) + stacksize);
  34
  =>      ABTD_thread_context_make(p_newctx, p_stacktop, stacksize, f_wrapper);
  36      p_newctx->f_thread = f_thread;
  37      p_newctx->p_arg = p_arg;
  38      ABTD_atomic_relaxed_store_thread_context_ptr(&p_newctx->p_link, p_link);
  39
  40      return abt_errno;
  41  }
```
The values shown are
```
Stack Frame
Function "ABTDI_thread_context_create":
  p_newctx:              0x2000024c7a10 -> (ABTD_thread_context)
  p_stack:               0x200000002000 -> 0x3abd017b3a9d01b1 (4232540853574107569)
  stacksize:             0x000020000013236c (35184373343084)
  p_arg:                 0x00000001 -> <Bad address: 0x00000001>
  f_thread:              0x00000000
  p_link:                0x00000000
Block "$b1":
  abt_errno:             0x00000000 (0)
  p_stacktop:            0x00003cb0 -> <Bad address: 0x00003cb0>
```
That ``p_stacktop`` value is shown correctly, and that is what leads to the segfault.  However, this is computed a couple lines up from:
```
p_stacktop = (void *)(((char *)p_stack) + stacksize);
```
And I don't yet see how ``p_stack=0x200000002000 + stacksize=0x000020000013236c`` turns into ``p_stacktop=0x00003cb0``.  Summing those two in a calculator, I get ``0x40000013436C``,even assuming that new leading bit gets chopped, the trailing part does not match.

It's hard to tell if those input values are shown correctly (due to optimization level messing with totalview), or they are shown correctly, and something strange happens in the casting and pointer arithmetic.

As the next step, let's try to hack the argobots code to see if we can printf the values once they've gone bad.

**Comment by adammoody** [comment]
I gave you a patch.  Grab it from me with the ``take`` command.  Place this patch file in the same directory as your bootstrap.sh script, then add a patch command into your bootstrap script as shown here:
```
./autogen.sh && CC=gcc ./configure CFLAGS="-g -O0" --prefix="$INSTALL_DIR" --enable-fast=O1
make clean
patch -p 1 < ../../context_create.patch
make V=1 VERBOSE=1 -j $make_nproc && make install
```
Then run bootstrap to build argobots again.

My goal here is to print those variables to stdout and then either have the server segfault or hang in case we can also attach with a debugger there.

**Comment by adammoody** [comment]
For future reference, this might be worth trying:
```
       -Og Optimize debugging experience.  -Og enables optimizations that do not interfere with debugging. It should be the optimization level of choice
           for the standard edit-compile-debug cycle, offering a reasonable level of optimization while maintaining fast compilation and a good debugging
           experience.
```

**Comment by adammoody** [comment]
Still working with the "-g O1" build for now...  Noting that the stacksize value ``(0x000020000013236c)`` looks a lot like other memory addresses, I was curious as to where it might be pointing.  Based on the ``/proc/<pid>/maps`` file for the server process, it would fall somewhere within an executable section of libmercury.so
```
200000120000-200000150000 r-xp 00000000 00:3a 1565731850                 /path/to/libmercury.so.2.0.0
```

**Comment by adammoody** [comment]
We should try setting the following on this problem, too:
```
export ABT_THREAD_STACKSIZE=256000
```
This should be set within the environment for the server process(es).

**Comment by adammoody** [comment]
Setting ``export ABT_THREAD_STACKSIZE=256000`` seems to be a work around for this issue, so this is likely due to stack overflow, as well.

**Comment by adammoody** [comment]
Closing with the hope that a small stack size was the issue.
